File: DotNetGotchas/CSharp/Synchronizing/SynchOnType/Bacteria.cs
//Bacteria.cs

using System;
using System.Threading;

namespace SynchOnType
{
	public class Bacteria
	{
		private static int bacteriaCount;

		private static void IncreaseCount()
		{
			Console.WriteLine(
				"IncreaseCount called by {0} at {1}",
				AppDomain.GetCurrentThreadId(),
				DateTime.Now.ToLongTimeString());

			bacteriaCount++;

			Thread.Sleep(2000);
			// Used for illustration purpose
		}

		public Bacteria()
		{
			lock(this)
			{
				IncreaseCount();
			}
		}
	}
}

File: DotNetGotchas/VBDN/Synchronizing/SynchOnType/Bacteria.vb
'Bacteria.vb

Imports System.Threading

Public Class Bacteria
    Private Shared bacteriaCount As Integer

    Private Shared Sub IncreaseCount()
        Console.WriteLine( _
            "IncreaseCount called by {0} at {1}", _
            AppDomain.GetCurrentThreadId(), _
            DateTime.Now.ToLongTimeString())

        bacteriaCount += 1

        Thread.Sleep(2000)
        ' Used for illustration purpose
    End Sub
    Public Sub New()
        SyncLock Me
            IncreaseCount()
        End SyncLock
    End Sub
End Class
